File: Data Structures & Algorithms/merge-intervals/submission-0.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        //edge case: no intervals
        if(intervals.empty()) return {};

        //sort intervals by start time
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> res;
        for(auto& interval : intervals){
            //if result is empty or no overlap with last interval
            if(res.empty()||res.back()[1] < interval[0]){
                res.push_back(interval);
            } //overlap case -> merger by extending end
            else{
                res.back()[1] = max(res.back()[1], interval[1]);
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/non-overlapping-intervals/submission-0.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        count = 0
        intervals.sort()
        prevEnd = intervals[0][1]
        i = 1
        while i < len(intervals):
            if intervals[i][0] >= prevEnd:
                prevEnd = intervals[i][1]
                i+=1
            else:
                count += 1
                prevEnd = min(intervals[i][1], prevEnd)
                i+=1
        return count

            